### PR TITLE
14.0 fix reconcile payment in locked period poma

### PR DESCRIPTION
**Pull request by poma-odoo**
[FIX] account: accounting date for moves happen in lock period should be chosen based on sequence reset period (month/year)

also fixes:
[FIX] account: caba reconciliation in lock period
[FIX] account: foreign currency reconciliation/unreconciliation in lock period raise error

The accounting date and the related warnings were in some cases not
being set properly. This change aims to improve that by displaying better
warnings in the view and updating the date on post instead of raising
user errors.

Back port of:
5883f0a84cd3cd2330cc5ff34261252dacd8c5b9
b5fe668ce3998c39d09b28318efc58b6eb2aed88

Task : [2853102](https://www.odoo.com/web#id=2853102&menu_id=3940&cids=1&action=333&active_id=967&model=project.task&view_type=form)
Task: [2793033](https://www.odoo.com/web#id=2793033&cids=1&menu_id=3940&action=4043&model=project.task&view_type=form)

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by h4818 — COMMENTED** [review]
Seems like there are other commits which should not be part of this PR?

**Comment by rafaelbn** [comment]
From https://github.com/odoo/odoo/pull/91580

@moduon MT-719

OPW-2856385

File: addons/account/models/account_partial_reconcile.py [hunk 1]
```diff
@@ -119,7 +119,7 @@ def unlink(self):
         # Reverse CABA entries.
         today = fields.Date.context_today(self)
         default_values_list = [{
-            'date': move.date if move.date > (move.company_id.period_lock_date or date.min) else today,
+            'date': move._get_accounting_date(move.date, True),
```
Comment (smetl) [inline addons/account/models/account_partial_reconcile.py:122]: I don't understand all this bunch of codes... This single line is fixing Bug 1.
Bug 2 is only a traceback..
Imo your diff should be 4-5 lines, not more

**Review by smetl — CHANGES_REQUESTED** [review]

File: addons/account/models/account_partial_reconcile.py (hunk 1 above)
Comment (poma-odoo) [reply addons/account/models/account_partial_reconcile.py:122]: Actually my fix is 3 lines, you can see it in my commit, the other part is the back port of VIN's patch for displaying lock dates to the user. The reason that I need other changes besides this is that there were other cases provided by TSB, I have added them in tests.

File: addons/account/models/account_full_reconcile.py [hunk 2]
```diff
@@ -29,7 +29,7 @@ def unlink(self):
         # Reverse all exchange moves at once.
         today = fields.Date.context_today(self)
         default_values_list = [{
-            'date': today,
```
Comment (h4818) [inline addons/account/models/account_full_reconcile.py:32]: today is no longer used

File: addons/account/models/account_full_reconcile.py [hunk 3]
```diff
@@ -29,7 +29,7 @@ def unlink(self):
         # Reverse all exchange moves at once.
         today = fields.Date.context_today(self)
```
Comment (Shide) [inline addons/account/models/account_full_reconcile.py:30]: Remove today from here too (based on https://github.com/odoo/odoo/pull/91245/files/68a8917cd161cff68d9ac42983a40c8e91654301#r899984046)

**Review by Shide — CHANGES_REQUESTED** [review]
Only the removal of the unused variable.

**Comment by poma-odoo** [comment]
The reason for using `in_unreconcile` is that we don't want to change the behavior globally in V13-15 because it can be considered as a breaking change. Though in master, we don't need this and can set accounting date as a standard feature.

**Comment by poma-odoo** [comment]
@fw-bot up to 15.0

**Comment by qdp-odoo** [comment]
> The reason for using `in_unreconcile` is that we don't want to change the behavior globally in V13-15 because it can be considered as a breaking change. Though in master, we don't need this and can set accounting date as a standard feature.

indeed, that's what I was about to say: this is a behavior change, so there shouldn't be a fix for that at all

**Comment by Shide** [comment]
The behaviour must change because it's a bug when creating the exchange
move on payment cancelling...

Before: "today"
Fixed: move.date

This should be the behaviour change.

**Comment by rafaelbn** [comment]
Hello @poma-odoo @h4818 @smetl ! 😄 

Do you understand that you have during months (may, june, july, august) a BUG in production of your Enterprise Customers just because you don't finish a review of code?

Now you have to think again about code, review again because surely you don't remenber it fresh like en may/june

I cannot understand why don't you finish review in the moment you have worked it.

- We make the best a Partner can do, open a very well explained OPW-2856385 (May 17th) with the PR which fix it https://github.com/odoo/odoo/pull/91580 that your coleguaes @william-andre @poma-odoo and @h4818 
- Here https://github.com/odoo/odoo/pull/91245#issuecomment-1169666293 we explain you crearly the main problem that our Enterprise Customer have EVERYDAY because it's a international company.
- In the OPW you have a really EASY 2 minutes VIDEO where you can see and reproduce this use case with you FIX in order to check if you are fixing it. OPW-2856385 from @moduon

This means you still have Odoo Task long time without solving:
- Task : [2853102](https://www.odoo.com/web#id=2853102&menu_id=3940&cids=1&action=333&active_id=967&model=project.task&view_type=form)
- Task: [2793033](https://www.odoo.com/web#id=2793033&cids=1&menu_id=3940&action=4043&model=project.task&view_type=form)

This is a OPW from may.

Thank you for your efforts and understanding.

Regards.

**Comment by guva-odoo** [comment]
Hello @smetl, is there a chance for this PR to be merged soon ?

**Comment by tsb-odoo** [comment]
Closed in favour #108618

**Comment by Shide** [comment]
Still not working, check out this comment: https://github.com/odoo/odoo/pull/108618#issuecomment-1369490490

**Pull request closed without merging**